### PR TITLE
Refactor AlgoliaWidget into several single method interfaces

### PR DESCRIPTION
**Pull request by PLNech**
This refactors `AlgoliaWidget` and `AlgoliaResultsListener` to improve DX and enforce [Interface Segregation](https://en.wikipedia.org/wiki/Interface_segregation_principle).

*Please focus your attention on cceb1ef, the other commits are either chores or minor changes.*

# Advantages
- Simpler architecture by removing the unnecessary `Widget` <> `Listener` distinction
- A widget does not need to implement all methods anymore (no more `onError` method stubs)

# Drawbacks
- [Widget detection](https://github.com/algolia/instantsearch-android/pull/39/files#diff-0ddf41769b716f6549384270cfc5fefdL292) is more complex (and has to ensure each widget is only processed once)
- [Widget processing](https://github.com/algolia/instantsearch-android/pull/39/files#diff-0ddf41769b716f6549384270cfc5fefdL307) is more verbose (and has code duplication, unless we want to use reflection)
- [Widget definition](https://github.com/algolia/instantsearch-android/pull/39/files#diff-c56bb82a5e91d21715bf32d5f233f5caL63) is a lot more verbose:
```java
/* Before */ public class Hits extends RecyclerView implements AlgoliaWidget { /* ... */ }
/* After */ public class Hits extends RecyclerView implements AlgoliaResultListener, AlgoliaErrorListener, AlgoliaSearcherListener { /* ... */ }

// It's even worse for an `AlgoliaFacetFilter`, which was a specialized `AlgoliaWidget`:
/* Before */ public abstract class Toggle extends AppCompatCheckBox implements AlgoliaFacetFilter { /* ... */ }
/* After */ public abstract class Toggle extends AppCompatCheckBox implements AlgoliaFacetFilter, AlgoliaResultListener, AlgoliaErrorListener, AlgoliaSearcherListener { /* ... */ }
```

@spinach: I'd love you to review this: if you think it improves the DX, happy to ship it in v1 🙂

### CODE REVIEW

**Comment by PLNech** [comment]
Deploy preview processing.

Built with commit cceb1ef276e8b431632f4e138c36478e0252a866

https://app.netlify.com/sites/instantsearch-android-preview/deploys/596799a8424ef26300a899da

File: instantsearch/src/main/java/com/algolia/instantsearch/model/AlgoliaSearcherListener.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.algolia.instantsearch.model;
+
+import android.support.annotation.NonNull;
+
+import com.algolia.instantsearch.helpers.Searcher;
+
+public interface AlgoliaSearcherListener {
```
Comment (spinach) [inline instantsearch/src/main/java/com/algolia/instantsearch/model/AlgoliaSearcherListener.java]: nit: You can some javadoc here explaining what is this interface for:
- for widgets who need a reference to a Searcher
- Mainly used for "input" widgets that will need to modify the state of the searcher

File: instantsearch/src/main/java/com/algolia/instantsearch/ui/InstantSearch.java [hunk 2]
```diff
@@ -289,53 +285,89 @@ private void processActivity(@NonNull final Activity activity) {
...
+        for (AlgoliaErrorListener listener : errorListeners) {
+            processErrorListener(rootView, listener, refinementAttributes);
+        }
+
+        // Register any AlgoliaSearcherListener
+        final List<AlgoliaSearcherListener> searcherListeners = LayoutViews.findByClass(rootView, AlgoliaSearcherListener.class);
+        for (AlgoliaSearcherListener listener : searcherListeners) {
+            processSearcherListener(rootView, listener, refinementAttributes);
         }
+
```
Comment (spinach) [inline instantsearch/src/main/java/com/algolia/instantsearch/ui/InstantSearch.java]: refactor nit: you can extract L290-L309 into a method processAllListeners(rootView, listener, refinementAttributes)

**Comment by spinach** [comment]
Looks good to me! I guess you're also working on adding the Refinable Listeners next 👍 

Concerning your drawbacks:

> Widget detection is more complex
That's correct, but we won't have too many widget interfaces so I think it's ok

> Widget processing is more verbose
This could potentially be solved if we introduce more interfaces and rethink a bit the architecture, but I say it's not worth it to over-engineer for just 3-4 process functions that are a few lines of code, especially that in the long run, we will probably not add more than that.

> Widget definition is a lot more verbose
For me this is subjective: sometimes more verbose means it gives more info and hides less "magic" from the developer, which I take it as an advantage :)

File: instantsearch/src/main/java/com/algolia/instantsearch/model/AlgoliaSearcherListener.java (hunk 1 above)
Comment (PLNech) [inline instantsearch/src/main/java/com/algolia/instantsearch/model/AlgoliaSearcherListener.java]: 👍  see 260ab50a.

File: instantsearch/src/main/java/com/algolia/instantsearch/ui/InstantSearch.java (hunk 2 above)
Comment (PLNech) [inline instantsearch/src/main/java/com/algolia/instantsearch/ui/InstantSearch.java]: Refactored (and simplified refinementAttributes handling) in c1090e34.

**Pull request merged**